### PR TITLE
[core] Minor optimization and thread annotation for store client

### PR DESCRIPTION
**Pull request by dentiny**
I'm reading through the codebase to get myself familiar, some minor optimization and thread annotations in this PR.

### CODE REVIEW

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -2451,9 +2451,8 @@ Status CoreWorker::CreatePlacementGroup(
            << ". It is probably "
               "because GCS server is dead or there's a high load there.";
     return Status::TimedOut(stream.str());
-  } else {
```
Comment (dentiny) [inline src/ray/core_worker/core_worker.cc:2454]: clang-tidy has an item for this, which encourages early return to reduce indentation.

Comment (dentiny) [reply src/ray/core_worker/core_worker.cc:2454]: Reference: https://clang.llvm.org/extra/clang-tidy/checks/readability/else-after-return.html

File: src/ray/gcs/store_client/in_memory_store_client.h [hunk 2]
```diff
@@ -88,9 +90,8 @@ class InMemoryStoreClient : public StoreClient {
   /// of the callback.
   instrumented_io_context &main_io_service_;
 
-  int job_id_ = 0;
+  /// Current job id, auto-increment when request next-id.
+  int job_id_ ABSL_GUARDED_BY(mutex_) = 0;
```
Comment (dentiny) [inline src/ray/gcs/store_client/in_memory_store_client.h:94]: Apply thread annotation to methods and data members.

**Review by dayshah — APPROVED** [review]

File: src/ray/gcs/store_client/in_memory_store_client.cc [hunk 3]
```diff
@@ -66,9 +65,9 @@ Status InMemoryStoreClient::AsyncGetAll(
     const std::string &table_name,
     const MapCallback<std::string, std::string> &callback) {
   RAY_CHECK(callback);
+  auto result = absl::flat_hash_map<std::string, std::string>();
```
[the remark this answers is not in the document]
Comment (dentiny) [reply src/ray/gcs/store_client/in_memory_store_client.cc:68]: Updated.

File: src/ray/gcs/store_client/in_memory_store_client.cc [hunk 4]
```diff
@@ -171,11 +171,10 @@ Status InMemoryStoreClient::AsyncExists(const std::string &table_name,
   auto table = GetOrCreateTable(table_name);
   absl::MutexLock lock(&(table->mutex_));
   bool result = table->records_.contains(key);
-  main_io_service_.post([result, callback]() mutable { callback(result); },
-                        "GcsInMemoryStore.Exists");
+  main_io_service_.post(
+      [result, callback = std::move(callback)]() mutable { callback(result); },
```
[the remark this answers is not in the document]
Comment (dentiny) [reply src/ray/gcs/store_client/in_memory_store_client.cc:176]: Updated.

File: src/ray/gcs/store_client/in_memory_store_client.cc [hunk 5]
```diff
@@ -138,28 +137,29 @@ std::shared_ptr<InMemoryStoreClient::InMemoryTable> InMemoryStoreClient::GetOrCr
   auto iter = tables_.find(table_name);
   if (iter != tables_.end()) {
     return iter->second;
-  } else {
-    auto table = std::make_shared<InMemoryTable>();
-    tables_[table_name] = table;
-    return table;
   }
+  auto table = std::make_shared<InMemoryTable>();
```
[the remark this answers is not in the document]
Comment (dentiny) [reply src/ray/gcs/store_client/in_memory_store_client.cc:142]: I think we're abusing shared pointer in our codebase.
In my experience, it's only necessary async or multi-threaded env, when lifecycle is impossible to justify.

But I will leave it here, it's out of this PR's scope. And I agree we should do some auditing on shared pointer usage.

[the remark this answers is not in the document]
Comment (dayshah) [reply src/ray/gcs/store_client/in_memory_store_client.cc:142]: agreed, way too many

**Review by jjyao — APPROVED** [review]

**Pull request merged**